src/Functions.tsx

```diff
@@ -150,9 +155,43 @@ const updateExistingRows = (
 
 const createCombinedRow = (
   averageCoordinates: Point,
-  rows: Types.row[]
+  rows: Types.row[],
+  index: number
 ): Types.combinedRow => {
-  return { averageCoordinates: averageCoordinates, rows: rows };
+  return {
+    averageCoordinates: averageCoordinates,
+    rows: rows,
+    index: index,
+    isMarkerVisible: getCombinedRowVisibility(index),
+  };
+};
+
+export const getCombinedRowVisibility = (index: number): Types.visibility => {
+  const elementId = getMarkerId(index);
+  const svgMarker = document.getElementById(elementId);
+  const isMarkerVisible = elementIsInViewport(svgMarker);
+  return isMarkerVisible;
+};
+
+export const getMarkerId = (index: number): string =>
+  Config.elementIds.Marker(index);
+
+const elementIsInViewport = (element: HTMLElement | null): Types.visibility => {
+  const rectangle = element?.getBoundingClientRect();
+  const container = document.getElementById("container");
+  const elementNotValid = !element || 1 !== element.nodeType;
+  const noSuchRectangle = rectangle === undefined;
+  const noSuchContainer = container === null;
+  if (elementNotValid || noSuchRectangle || noSuchContainer) {
```

Comment:
You can leverage truthiness here a bit. This made it so I didn't have to do those casts on 188 and 189
```suggestion
  if (elementNotValid || !rectangle || !container) {
```